File: src/net/dinkla/raytracer/brdf/PerfectSpecular.java
package net.dinkla.raytracer.brdf;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.math.Normal;
import net.dinkla.raytracer.math.Vector3D;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 24.04.2010
 * Time: 11:25:05
 * To change this template use File | Settings | File Templates.
 */
public class PerfectSpecular<C extends Color> extends BRDF<C> {

    public double kr;
    public C cr;

    public PerfectSpecular() {
        super();
        kr = 1.0;
        cr = (C) C.WHITE;
    }

    @Override
    public C f(final Shade sr, final Vector3D wo, final Vector3D wi){
        throw new RuntimeException("PerfectSpecular.f");
        // Im C-Code Black
    }

    @Override
    public Sample sampleF(Shade sr, Vector3D wo) {
        assert null != cr;
        final Sample result = new Sample();
        Normal normal = sr.getNormal();
        final double nDotWo = normal.dot(wo);
        result.wi = wo.negate().plus(sr.getNormal().mult(2.0 * nDotWo));
        final double nDotWi = normal.dot(result.wi);
        result.color = (C) cr.mult(kr / Math.abs(nDotWi));
        return result;
    }

    @Override
    public C rho(Shade sr, Vector3D wo) {
        throw new RuntimeException("PerfectSpecular.rho");
    }

}

File: src/main/kotlin/net/dinkla/raytracer/brdf/PerfectSpecular.kt
package net.dinkla.raytracer.brdf

import net.dinkla.raytracer.brdf.BRDF.Sample
import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.math.Vector3D
import kotlin.math.abs
import net.dinkla.raytracer.interfaces.hash

class PerfectSpecular(var kr: Double = 1.0, var cr: Color = Color.WHITE) : BRDF {

    override fun f(sr: Shade, wo: Vector3D, wi: Vector3D): Color {
        throw RuntimeException("PerfectSpecular.f")
        // TODO Im C-Code Black
    }

    override fun sampleF(sr: Shade, wo: Vector3D): Sample {
        val normal = sr.normal
        val nDotWo = normal dot wo
        val wi = -wo + (sr.normal * (2.0 * nDotWo))
        val nDotWi = normal dot wi
        return Sample(wi = wi, color = cr * (kr / abs(nDotWi)), pdf = 1.0)
    }

    override fun rho(sr: Shade, wo: Vector3D): Color {
        throw RuntimeException("PerfectSpecular.rho")
    }

    override fun equals(other: Any?): Boolean {
        if (other != null && other is PerfectSpecular) {
            return kr == other.kr && cr == other.cr
        }
        return false
    }

    override fun hashCode(): Int = hash(kr, cr)

    override fun toString() = "PerfectSpecular($kr, $cr)"
}
